src/uschedulecp.1.in
.TH uschedulecp 1 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uschedulecp - copy a uschedule command
.LINE
.section SYNOPSIS
\fBuschedulecp\fR [-.] [-d DIR] OLD-ID [NEW-ID]
.section DESCRIPTION
\fBuschedulecp\fR creates a new job by copying an old one. The new
jobs has the identifier \fINEW-ID\fR or, if that is not given, will
be assigned an unique identifier.
.section OPTIONS
.options
.option "-., --dot-as-home"
The current working directory will be used instead of $HOME.
.option "-d, --dir=DIR"
Put the new job into \fIDIR\fR. The default is \fI~/.uschedule/\fR.
.section CAVEAT
Note that this just creates the command to be run, it doesn't 
uschedule the job to be run at any time. Use the
.manlink uschedule.html uschedule(1)
command for this.
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedule.html uschedule(1) ,
.manlink uschedulecmd.html uschedulecmd(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uschedulecp.c
#include "uogetopt.h"
#include "scheduled.h"
#include "open.h"
#include "attributes.h"
#include "stralloc.h"
#include "bailout.h"
#include "buffer.h"
#include "error.h"
#include <unistd.h>
#include <sys/stat.h>

static const char *oldid;
static const char *newid;
static const char *o_dir;
static int o_dot_as_home;

static uogetopt2 myopts[]={
{'.',"dot-as-home",uogo_flag,UOGO_NOARG,&o_dot_as_home, 1 ,
"Use current directory instead of $HOME.",0,0},
{'d',"dir",uogo_string,0,&o_dir, 0 ,
"Jobdirectory.",
"Jobs will be read from this directory.","DIR"},
{0,0,0,0,0,0,0,0,0}
};

static void die_write(void) attribute_noreturn;
static void die_write_temp(void) attribute_noreturn;
static void die_fsync(void) attribute_noreturn;
static void die_fchmod(void) attribute_noreturn;

static void die_write(void)  {xbailout(111,errno,"failed to write",0,0,0);}
static void die_write_temp(void)
{ xbailout(111,errno,"failed to write temporary file",0,0,0); }
static void die_fsync(void)
{ xbailout(111,errno,"failed to fsync temporary file",0,0,0); }
static void die_fchmod(void)
{ xbailout(111,errno,"failed to fchmod temporary file",0,0,0); }

static void 
cp1(int rfd, int wfd)
{
	buffer i;
	char ispace[512];
	buffer o;
	char ospace[512];
	
	buffer_init(&i,(buffer_op) read,rfd,ispace,sizeof(ispace)); 
	buffer_init(&o,(buffer_op) write,wfd,ospace,sizeof(ospace)); 
	switch (buffer_copy(&o,&i)) {
	case 0: break;
	case -2: xbailout(111,errno,"failed to read input",0,0,0);
	case -3: die_write_temp();
	default: xbailout(100,errno,"unexpected errorcode from buffer_copy",0,0,0);
	}

	if (-1==buffer_flush(&o)) die_write_temp();
	if (-1==fchmod(wfd,0755)) die_fchmod();
	if (-1==fsync(wfd)) die_fsync();
}

static void
copyit(int rfd, int wfd)
{
	static stralloc sa;

	cp1(rfd,wfd);
	close(wfd);

	if (!stralloc_copys(&sa,oldid)) oom();
	if (!stralloc_cats(&sa,".run")) oom();
	if (!stralloc_0(&sa)) oom();
	
	close(rfd);
	rfd=open_read(sa.s);
	if (-1==rfd) {
		int e=errno;
		if (errno==error_noent)
			return; /* <-------- point of return if no .run file exists */
		unlink(newid);
		xbailout(111,e,"failed to open_read ",sa.s,0,0);
	}
	if (!stralloc_copys(&sa,newid)) oom();
	if (!stralloc_cats(&sa,".run")) oom();
	if (!stralloc_0(&sa)) oom();

	wfd=open_excl(sa.s);
	if (-1==wfd) {
		int e=errno;
		unlink(newid);
		xbailout(111,e,"failed to open_excl ",sa.s,0,0);
	}
	cp1(rfd,wfd);
	close(wfd);

}

static uogetopt_env myoptenv={
"uschedulecp",PACKAGE,VERSION,
"uschedulecp [options] OLD-ID [NEW-ID]",
"This program creates a new uschedule job by copying an old one.\n",
"long",
"Report bugs to uschedule@lists.ohse.de",
2,3,0,0,uogetopt_out,myopts
};


int main(int argc, char **argv) 
{
	struct tai now;
	int wfd=-1;
	int rfd;
	int prt=0;
	tai_now(&now);
	bailout_progname(argv[0]);
	flag_bailout_fatal_begin=3;
	uogo_posixmode=1;
	myoptenv.program=flag_bailout_log_name;
	uogetopt_parse(&myoptenv,&argc,argv);

	change_dir(0,o_dir,o_dot_as_home);
	if (-1==chdir(IDDIR))
		xbailout(111,errno,"failed to chdir to ",IDDIR, " directory",0);
	oldid=argv[1];
	newid=argv[2];

	rfd=open_read(oldid);
	if (-1==rfd) 
		xbailout(111,errno,"failed to open job ",oldid, " for reading",0);
	if (!newid) {
		static stralloc sa;
		wfd=make_id(&sa);
		newid=sa.s;
		prt=1;
	} else {
		check_id(newid);
		wfd=open_excl(newid);
		if (-1==wfd)
			xbailout(111,errno,"failed to open_excl ",newid,0,0);
	}
	copyit(rfd,wfd);
	if (prt) {
		if (-1==buffer_puts(buffer_1,"The id of the new job is ")) die_write();
		if (-1==buffer_puts(buffer_1,newid)) die_write();
		if (-1==buffer_puts(buffer_1,"\n")) die_write();
		if (-1==buffer_flush(buffer_1)) die_write();
	}        
	return 0;
}
